### PR TITLE
Updated mode operation for  Digital APIs Part 2

### PR DESCRIPTION
**Pull request by nitamil**
### What does this Pull Request accomplish?

Adds UpdateMode support to digital configuration extension methods.

- ConfigureTimeSetPeriod
- ConfigureTimeSetDriveEdges
- ConfigureTimeSetEdge

The parameter defaults to UpdateMode.Deferred to preserve existing behavior. The selected update mode is applied through a new internal ApplyUpdateMode extension method, supporting:

- UpdateMode.Deferred
- UpdateMode.Commit
- UpdateMode.Immediate

The following methods now accept an optional updateMode parameter:

### Why should this Pull Request be merged?

This change provides callers with explicit control over when digital configuration changes are applied to the instrument.

It also makes the digital configuration APIs consistent with other instrument abstraction APIs that already support update modes. Existing applications remain compatible because the default behavior continues to be deferred updates.

### What testing has been done?
- Added unit tests covering all updated configuration methods.
- Verified behavior for:

    - UpdateMode.Deferred
    - UpdateMode.Commit
    - UpdateMode.Immediate

- Added coverage for the update-mode handling implemented by ApplyUpdateMode.
- Confirmed that the existing configuration behavior remains unchanged when no update mode is specified.

ConfigureTimeSetDriverEdges case:
image

ConfigureTimeSetDriverEdgesWithTwoDriveData case:
image

ConfigureTimeSetPeriodWithoutSpecifyingPins case:
image

ConfigureEdgeWithSingleValue case:
image

ConfigureEdgeWithSiteSpecificValues case:
image

ConfigureEdgeWithChannelSpecificValues case:
image

### CODE REVIEW

File: SemiconductorTestLibrary.Extensions/source/InstrumentAbstraction/Digital/LevelsAndTiming.cs [hunk 1]
```diff
@@ -183,12 +183,14 @@ public static void ConfigureTimeSetCompareEdgesStrobe(this DigitalSessionsBundle
         /// <param name="sessionsBundle">The <see cref="DigitalSessionsBundle"/> object.</param>
         /// <param name="timeSet">The name of the time set.</param>
         /// <param name="period">The period to configure.</param>
-        public static void ConfigureTimeSetPeriod(this DigitalSessionsBundle sessionsBundle, string timeSet, double period)
+        /// <param name="updateMode">Specifies when the configured settings are applied: <see cref="UpdateMode.Deferred"/> applies on the next burst or initiate operation, <see cref="UpdateMode.Commit"/> commits immediately, and <see cref="UpdateMode.Immediate"/> initiates immediately.</param>
+        public static void ConfigureTimeSetPeriod(this DigitalSessionsBundle sessionsBundle, string timeSet, double period, UpdateMode updateMode = UpdateMode.Deferred)
         {
```
[the remark this answers is not in the document]
Comment (nitamil) [reply SemiconductorTestLibrary.Extensions/source/InstrumentAbstraction/Digital/LevelsAndTiming.cs:218]: won't fix

File: SemiconductorTestLibrary.Extensions/tests/Unit/InstrumentAbstraction/Digital/LevelsAndTimingTests.cs [hunk 2]
```diff
@@ -330,6 +330,21 @@ public void SessionsInitialized_ConfigureTimeSetDriverEdges_ValueCorrectlySet(st
             });
         }
 
+        [Theory]
+        [InlineData(UpdateMode.Commit)]
+        [InlineData(UpdateMode.Immediate)]
+        [InlineData(UpdateMode.Deferred)]
+        public void SessionsInitialized_ConfigureTimeSetDriverEdges_UpdateModeSetCorrectly(UpdateMode updateMode)
```
Comment (gmrsekar) [inline SemiconductorTestLibrary.Extensions/tests/Unit/InstrumentAbstraction/Digital/LevelsAndTimingTests.cs]: Don't we want to include "WithUpdateMode" Suffix in the Act part of test name?

Comment (nitamil) [reply SemiconductorTestLibrary.Extensions/tests/Unit/InstrumentAbstraction/Digital/LevelsAndTimingTests.cs]: Fixed it.

**Review by gmrsekar — APPROVED** [review]

File: SemiconductorTestLibrary.Extensions/tests/Unit/InstrumentAbstraction/Digital/LevelsAndTimingTests.cs [hunk 3]
```diff
@@ -1347,6 +1377,21 @@ public void SessionsInitialized_ConfigureTimeSetPeriodWithoutSpecifyingPins_Valu
...
+        [InlineData(UpdateMode.Deferred)]
+        [InlineData(UpdateMode.Commit)]
+        [InlineData(UpdateMode.Immediate)]
+        public void SessionsInitialized_ConfigureTimeSetPeriodWithoutSpecifyingPinsAndUpdateMode_UpdateModeSetCorrectly(UpdateMode updateMode)
+        {
+            var sessionManager = InitializeSessionsAndCreateSessionManager("Mixed Signal Tests.pinmap", "Mixed Signal Tests.digiproj");
+
+            var sessionsBundle = sessionManager.Digital();
+            sessionsBundle.ConfigurePattern("TX_50_Duty_Cycle");
+            sessionsBundle.ConfigureTimeSetPeriod("TS_SW", 5e-6, updateMode);
```
Comment (NikhilSivadas-NIB) [inline SemiconductorTestLibrary.Extensions/tests/Unit/InstrumentAbstraction/Digital/LevelsAndTimingTests.cs:1408]: For my understanding, any specific reason why ConfigureTimeSetPeriod focusing only on Bundle without specifying pins? previous test follows the same, is that the general use case or requirement for this method ?

File: SemiconductorTestLibrary.Extensions/tests/Unit/InstrumentAbstraction/Digital/LevelsAndTimingTests.cs [hunk 4]
```diff
@@ -1347,6 +1377,21 @@ public void SessionsInitialized_ConfigureTimeSetPeriodWithoutSpecifyingPins_Valu
             });
         }
 
+        [Theory]
+        [InlineData(UpdateMode.Deferred)]
+        [InlineData(UpdateMode.Commit)]
+        [InlineData(UpdateMode.Immediate)]
+        public void SessionsInitialized_ConfigureTimeSetPeriodWithoutSpecifyingPinsAndUpdateMode_UpdateModeSetCorrectly(UpdateMode updateMode)
```
Comment (NikhilSivadas-NIB) [inline SemiconductorTestLibrary.Extensions/tests/Unit/InstrumentAbstraction/Digital/LevelsAndTimingTests.cs]: Doesn't the name sound like ConfigureTimeSetPeriod is called without pins and update mode?

How about "ConfigureTimeSetPeriodWithUpdateModeAndWithoutPins" if we want to mention both, otherwise just"ConfigureTimeSetPeriodWithUpdateMode", thoughts?

File: SemiconductorTestLibrary.Extensions/tests/Unit/InstrumentAbstraction/Digital/LevelsAndTimingTests.cs [hunk 5]
```diff
@@ -1347,6 +1377,21 @@ public void SessionsInitialized_ConfigureTimeSetPeriodWithoutSpecifyingPins_Valu
...
+        {
+            var sessionManager = InitializeSessionsAndCreateSessionManager("Mixed Signal Tests.pinmap", "Mixed Signal Tests.digiproj");
+
+            var sessionsBundle = sessionManager.Digital();
+            sessionsBundle.ConfigurePattern("TX_50_Duty_Cycle");
```
Comment (NikhilSivadas-NIB) [inline SemiconductorTestLibrary.Extensions/tests/Unit/InstrumentAbstraction/Digital/LevelsAndTimingTests.cs:1407]: some tests in this PR and previous PR added  this call "sessionsBundle.ConfigurePattern" as part of arrange section and some other tests clubbed with Act section, can we keep it consistent?

File: SemiconductorTestLibrary.Extensions/tests/Unit/InstrumentAbstraction/Digital/LevelsAndTimingTests.cs (hunk 3 above)
Comment (nitamil) [reply SemiconductorTestLibrary.Extensions/tests/Unit/InstrumentAbstraction/Digital/LevelsAndTimingTests.cs:1408]: yes. Typically the TimeSetPeriod is for all the pins.

File: SemiconductorTestLibrary.Extensions/tests/Unit/InstrumentAbstraction/Digital/LevelsAndTimingTests.cs (hunk 5 above)
Comment (nitamil) [reply SemiconductorTestLibrary.Extensions/tests/Unit/InstrumentAbstraction/Digital/LevelsAndTimingTests.cs:1407]: Sure. I will make it consistent.

File: SemiconductorTestLibrary.Extensions/tests/Unit/InstrumentAbstraction/Digital/LevelsAndTimingTests.cs (hunk 4 above)
Comment (nitamil) [reply SemiconductorTestLibrary.Extensions/tests/Unit/InstrumentAbstraction/Digital/LevelsAndTimingTests.cs]: `ConfigureTimeSetPeriodWithUpdateMode` going with this name.

File: SemiconductorTestLibrary.Extensions/tests/Unit/InstrumentAbstraction/Digital/LevelsAndTimingTests.cs [hunk 6]
```diff
@@ -1484,6 +1529,21 @@ public void SessionsInitialized_ConfigureEdgeWithSingleValue_ValueCorrectlySet(s
...
+        [InlineData(UpdateMode.Deferred)]
+        [InlineData(UpdateMode.Commit)]
+        [InlineData(UpdateMode.Immediate)]
+        public void SessionsInitialized_ConfigureTimeSetEdgeWithSingleValueAndUpdateMode_UpdateModeSetCorrectly(UpdateMode updateMode)
+        {
+            var sessionManager = InitializeSessionsAndCreateSessionManager("TwoDevicesWorkForTwoSitesSeparately.pinmap", "TwoDevicesWorkForTwoSitesSeparately.digiproj");
+            var sessionsBundle = sessionManager.Digital(new string[] { "C0", "C1" });
+            sessionsBundle.ConfigurePattern("TX_50_Duty_Cycle");
+
+            sessionsBundle.ConfigureTimeSetEdge("TS", TimeSetEdge.CompareStrobe, 5e-6, updateMode);
```
Comment (NikhilSivadas-NIB) [inline SemiconductorTestLibrary.Extensions/tests/Unit/InstrumentAbstraction/Digital/LevelsAndTimingTests.cs:1565]: Some of the existing configure tests of this modified APIs has ValueCorrectSet as part of Assert, similarly can we add one more test for such methods which could ensure values also set when different updateMode passed, Applicable to all existing configure methods and overloads that do value check a spart of Assert

For example:  SessionsInitialized_ConfigureTimeSetEdgeWithSingleValueAndUpdateMode_ValueCorrectlySet

**Review by NikhilSivadas-NIB — COMMENTED** [review]
Can you check the comments ?

File: SemiconductorTestLibrary.Extensions/tests/Unit/InstrumentAbstraction/Digital/LevelsAndTimingTests.cs (hunk 5 above)
Comment (nitamil) [reply SemiconductorTestLibrary.Extensions/tests/Unit/InstrumentAbstraction/Digital/LevelsAndTimingTests.cs:1407]: Updated.

File: SemiconductorTestLibrary.Extensions/tests/Unit/InstrumentAbstraction/Digital/LevelsAndTimingTests.cs (hunk 6 above)
Comment (nitamil) [reply SemiconductorTestLibrary.Extensions/tests/Unit/InstrumentAbstraction/Digital/LevelsAndTimingTests.cs:1565]: We have existing tests that validates the values are set correctly. We need to ensure only the additional functionality that we are adding that is the UpdateMode operation.

Comment (NikhilSivadas-NIB) [reply SemiconductorTestLibrary.Extensions/tests/Unit/InstrumentAbstraction/Digital/LevelsAndTimingTests.cs:1565]: My thought was that the existing test cases already verify the values are set correctly when UpdateMode isn't explicitly specified (default value). Since this change introduces support for different UpdateMode values, could we add a test to verify that values are still set correctly when a non default UpdateMode is passed explicitly?

For example :
```
[InlineData(UpdateMode.Commit)]
[InlineData(UpdateMode.Immediate)]
```

**Review by NikhilSivadas-NIB — COMMENTED** [review]
Can you check the reply?

**Comment by nitamil** [comment]
> Can you check the reply?

Fixed the review comments.

File: SemiconductorTestLibrary.Extensions/tests/Unit/InstrumentAbstraction/Digital/LevelsAndTimingTests.cs (hunk 6 above)
Comment (nitamil) [reply SemiconductorTestLibrary.Extensions/tests/Unit/InstrumentAbstraction/Digital/LevelsAndTimingTests.cs:1565]: Commit or Initiate is the last action. That will not affect the settings that are already stored in the session. 

I don't mind adding the logic but that will make the existing test case as duplicate (for default).

Comment (nitamil) [reply SemiconductorTestLibrary.Extensions/tests/Unit/InstrumentAbstraction/Digital/LevelsAndTimingTests.cs:1565]: Added the data validation as well.

Comment (NikhilSivadas-NIB) [reply SemiconductorTestLibrary.Extensions/tests/Unit/InstrumentAbstraction/Digital/LevelsAndTimingTests.cs:1565]: As we have combined it instead of adding new test, can we rename the Assert part of the tests to _ValueAndUpdateModeSetCorrectly?

**Review by NikhilSivadas-NIB — APPROVED** [review]
Approved with one comment
